src/rust/cryptography-x509-validation/src/certificate.rs

```diff
@@ -0,0 +1,109 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+//! Validation-specific certificate functionality.
+
+use cryptography_x509::certificate::Certificate;
+
+use crate::ops::CryptoOps;
+
+// TODO: Remove these attributes once we start using these helpers.
+#[allow(dead_code)]
+pub(crate) fn cert_is_self_issued(cert: &Certificate<'_>) -> bool {
+    cert.issuer() == cert.subject()
+}
+
+#[allow(dead_code)]
+pub(crate) fn cert_is_self_signed<B: CryptoOps>(cert: &Certificate<'_>, ops: &B) -> bool {
+    match ops.public_key(cert) {
+        Ok(pk) => cert_is_self_issued(cert) && ops.verify_signed_by(cert, pk).is_ok(),
+        Err(_) => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::certificate::Certificate;
+    use crate::ops::tests::{cert, v1_cert_pem, NullOps};
+    use crate::ops::CryptoOps;
+
+    use super::{cert_is_self_issued, cert_is_self_signed};
+
+    #[test]
+    fn test_certificate_v1() {
+        let cert_pem = v1_cert_pem();
+        let cert = cert(&cert_pem);
+        let ops = NullOps {};
+
+        assert!(!cert_is_self_issued(&cert));
+        assert!(!cert_is_self_signed(&cert, &ops));
+    }
+
+    fn ca_pem() -> pem::Pem {
+        // From vectors/cryptography_vectors/x509/custom/ca/ca.pem
+        pem::parse(
+            "-----BEGIN CERTIFICATE-----
+MIIBUTCB96ADAgECAgIDCTAKBggqhkjOPQQDAjAnMQswCQYDVQQGEwJVUzEYMBYG
+A1UEAwwPY3J5cHRvZ3JhcGh5IENBMB4XDTE3MDEwMTEyMDEwMFoXDTM4MTIzMTA4
+MzAwMFowJzELMAkGA1UEBhMCVVMxGDAWBgNVBAMMD2NyeXB0b2dyYXBoeSBDQTBZ
+MBMGByqGSM49AgEGCCqGSM49AwEHA0IABBj/z7v5Obj13cPuwECLBnUGq0/N2CxS
+JE4f4BBGZ7VfFblivTvPDG++Gve0oQ+0uctuhrNQ+WxRv8GC177F+QWjEzARMA8G
+A1UdEwEB/wQFMAMBAf8wCgYIKoZIzj0EAwIDSQAwRgIhANES742XWm64tkGnz8Dn
+pG6u2lHkZFQr3oaVvPcemvlbAiEA0WGGzmYx5C9UvfXIK7NEziT4pQtyESE0uRVK
+Xw4nMqk=
+-----END CERTIFICATE-----",
+        )
+        .unwrap()
+    }
```

Comment:
Nitpick: maybe put this as a `pub(crate)` helper under `ops::tests`, like `v1_cert_pem`? But not strong opinion.